### PR TITLE
Add vk_ext_subgroup_size_control

### PR DESCRIPTION
**Pull request by mmerecki**

### CODE REVIEW

**Comment by mmerecki** [comment]
@alan-baker could you take a look, please?

File: samples/config_helper_vulkan.cc [hunk 1]
```diff
@@ -609,6 +612,27 @@ std::string deviceTypeToName(VkPhysicalDeviceType type) {
   return "unknown";
 }
 
+std::string stageFlagBitsToNames(const VkShaderStageFlags bits) {
+
```
Comment (dj2) [inline samples/config_helper_vulkan.cc:616]: nit: remove blank line

File: samples/config_helper_vulkan.cc [hunk 2]
```diff
@@ -609,6 +612,27 @@ std::string deviceTypeToName(VkPhysicalDeviceType type) {
...
+    std::string result;
```
Comment (dj2) [inline samples/config_helper_vulkan.cc:617]: This should be an `std::ostringstream` and move the declaration down to line 626

File: samples/config_helper_vulkan.cc [hunk 3]
```diff
@@ -609,6 +612,27 @@ std::string deviceTypeToName(VkPhysicalDeviceType type) {
...
+        std::make_pair(VK_SHADER_STAGE_TESSELLATION_EVALUATION_BIT, "tesse"),
+        std::make_pair(VK_SHADER_STAGE_GEOMETRY_BIT, "geom"),
+        std::make_pair(VK_SHADER_STAGE_FRAGMENT_BIT, "frag"),
+        std::make_pair(VK_SHADER_STAGE_COMPUTE_BIT, "comp")
+    };
+    std::string separator;
+    for (const auto& stage : stages) {
+        if ((bits & stage.first) != 0) {
+            result += separator + stage.second;
+            separator = ", ";
```
Comment (dj2) [inline samples/config_helper_vulkan.cc:630]: I'd prefer if this was a boolean to track if this is the first one emitted or not and add the ', ' if it isn't.

File: samples/config_helper_vulkan.cc [hunk 4]
```diff
@@ -1070,6 +1122,12 @@ void ConfigHelperVulkan::DumpPhysicalDeviceInfo() {
   if (vkGetPhysicalDeviceProperties2KHR) {
     std::cout << "  driverName: " << driver_properties.driverName << std::endl;
     std::cout << "  driverInfo: " << driver_properties.driverInfo << std::endl;
+    if (supports_subgroup_size_control_) {
+        std::cout << "  minSubgroupSize: " << subgroup_size_control_properties.minSubgroupSize << std::endl;
```
Comment (dj2) [inline samples/config_helper_vulkan.cc:1126]: These seem longer then 80 chars, there is a `./tools/format` script which will run `clang-format` over all the files.

File: src/pipeline.cc [hunk 5]
```diff
@@ -149,6 +155,84 @@ Result Pipeline::SetShaderCompileOptions(const Shader* shader,
...
+{
+    const bool isPow2 = (subgroupSize & (subgroupSize - 1)) == 0;
+    if (subgroupSize == 0 || subgroupSize > 128 || !isPow2) {
+        return Result("invalid required subgroup size " +
+            std::to_string(subgroupSize) +
+            " specified for shader name " +
+            shader->GetName());
+    }
+    const ShaderInfo::RequiredSubgroupSizeSetting subgroupSizeSetting =
+        static_cast<ShaderInfo::RequiredSubgroupSizeSetting>(subgroupSize);
```
Comment (dj2) [inline src/pipeline.cc:186]: I'd prefer this to be done a different way. Maybe we store a flag if it's numeric, max or min and then a second value which stores the numeric.

File: src/amberscript/parser_subgroup_size_control_test.cc [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+
+TEST_F(AmberScriptParserTest, SubgroupSizeControlRequiredSubgroupSizeInvalidSubgroupSize3) {
+    std::string in = R"(
+DEVICE_FEATURE SubgroupSizeControl.subgroupSizeControl
+SHADER compute test_shader GLSL
+# GLSL
+END
+PIPELINE compute pipeline
+  ATTACH test_shader
+  REQUIRED_SUBGROUP_SIZE test_shader 129
```
Comment (dj2) [inline src/amberscript/parser_subgroup_size_control_test.cc:215]: Make the 129 a power of 2 > 128

File: src/pipeline.cc [hunk 7]
```diff
@@ -149,6 +155,84 @@ Result Pipeline::SetShaderCompileOptions(const Shader* shader,
...
+  }
+
+  return Result("unknown shader specified for required subgroup size: " +
+                shader->GetName());
+}
+
+Result Pipeline::SetShaderRequiredSubgroupSize(const Shader* shader,
+    const uint32_t subgroupSize)
+{
+    const bool isPow2 = (subgroupSize & (subgroupSize - 1)) == 0;
```
Comment (dj2) [inline src/pipeline.cc:178]: If subgroupSize is 0 this is going underflow the `subgroupSize - 1`

File: src/pipeline.h [hunk 8]
```diff
@@ -63,6 +63,42 @@ class Pipeline {
       return compile_options_;
     }
 
+    enum class RequiredSubgroupSizeSetting : uint32_t {
+        NOT_SET = 0,
```
Comment (dj2) [inline src/pipeline.h:67]: Enum entries should be named as:

```
kNotSet
kSetTo1
...
```

File: src/script.h [hunk 9]
```diff
@@ -160,6 +160,14 @@ class Script : public RecipeImpl {
     engine_info_.required_features.push_back(feature);
   }
 
+  /// Checks is |feature| is in required features
```
Comment (dj2) [inline src/script.h:163]: nit: s/Checks is/Checks if/

File: src/script.h [hunk 10]
```diff
@@ -160,6 +160,14 @@ class Script : public RecipeImpl {
...
+  bool IsRequiredFeature(const std::string& feature) const {
+      return std::find(
+          engine_info_.required_features.cbegin(),
```
Comment (dj2) [inline src/script.h:166]: Why not just `begin` and `end`? Why `cbegin` and `cend`?

File: src/vulkan/device.cc [hunk 11]
```diff
@@ -390,16 +393,21 @@ Result Device::Initialize(
...
     const VkPhysicalDeviceFeatures& available_features,
     const VkPhysicalDeviceFeatures2KHR& available_features2,
     const std::vector<std::string>& available_extensions) {
   Result r = LoadVulkanPointers(getInstanceProcAddr, delegate);
   if (!r.IsSuccess())
     return r;
 
-  bool use_physical_device_features_2 = false;
-  for (auto& ext : required_extensions) {
+  const bool isVulkan11Plus =
```
Comment (dj2) [inline src/vulkan/device.cc:405]: Please match local naming conventions `is_vulkan_1_1_or_higher`

File: src/vulkan/device.cc [hunk 12]
```diff
@@ -390,16 +393,21 @@ Result Device::Initialize(
...
+      (VK_VERSION_MAJOR(physical_device_properties_.apiVersion) >= 1 &&
+       VK_VERSION_MINOR(physical_device_properties_.apiVersion) >= 1);
```
Comment (dj2) [inline src/vulkan/device.cc:407]: This isn't right. This would fail for version `2.0`

File: src/vulkan/device.cc [hunk 13]
```diff
@@ -390,16 +393,21 @@ Result Device::Initialize(
...
+
+  bool use_physical_device_features_2 = isVulkan11Plus;
```
Comment (dj2) [inline src/vulkan/device.cc:409]: This change seems unrelated to the change in progress, can it be moved to its own change.

File: src/vulkan/device.cc [hunk 14]
```diff
@@ -547,6 +574,26 @@ Result Device::Initialize(
   ptrs_.vkGetPhysicalDeviceMemoryProperties(physical_device_,
                                             &physical_memory_properties_);
 
+  subgroup_size_control_properties_ = {};
+  if (use_physical_device_features_2 &&
+      std::find(
```
Comment (dj2) [inline src/vulkan/device.cc:579]: Please move the find outside of the if conditional.

File: src/vulkan/device.cc [hunk 15]
```diff
@@ -547,6 +574,26 @@ Result Device::Initialize(
...
+          required_features.begin(),
+          required_features.end(),
+          kSubgroupSizeControl) != required_features.end()) {
+
+      PFN_vkGetPhysicalDeviceProperties2KHR vkGetPhysicalDeviceProperties2KHR =
```
Comment (dj2) [inline src/vulkan/device.cc:584]: This should use the provided `available_features2` instead of getting the features again.

File: src/vulkan/device.cc [hunk 16]
```diff
@@ -1010,5 +1057,54 @@ VkFormat Device::GetVkFormat(const Format& format) const {
...
+        break;
+    case kShaderTypeTessellationEvaluation:
+        stage = VK_SHADER_STAGE_TESSELLATION_EVALUATION_BIT;
+        break;
+    case kShaderTypeCompute:
+        stage = VK_SHADER_STAGE_COMPUTE_BIT;
+        break;
+    default:
+    case kShaderTypeMulti:
+        is_supported = false;
```
Comment (dj2) [inline src/vulkan/device.cc:1086]: Just `return false`

File: src/vulkan/device.cc [hunk 17]
```diff
@@ -1010,5 +1057,54 @@ VkFormat Device::GetVkFormat(const Format& format) const {
...
+        break;
+    }
+    if ((stage & subgroup_size_control_properties_.requiredSubgroupSizeStages) == 0) {
+        is_supported = false;
```
Comment (dj2) [inline src/vulkan/device.cc:1090]: These can both just `return false`. Since the if() then returns the `else if()` can just be an `if()`

File: src/vulkan/engine_vulkan.cc [hunk 18]
```diff
@@ -298,6 +299,38 @@ Result EngineVulkan::SetShader(amber::Pipeline* pipeline,
     if (shader_info.GetShaderType() != type)
       continue;
 
+    const amber::Pipeline::ShaderInfo::RequiredSubgroupSizeSetting required_subgroup_size =
```
Comment (dj2) [inline src/vulkan/engine_vulkan.cc:302]: `const auto required_subgroup_size`

File: src/vulkan/engine_vulkan.cc [hunk 19]
```diff
@@ -350,6 +384,18 @@ Result EngineVulkan::GetVkShaderStageInfo(
       stage_info[stage_count].pSpecializationInfo =
           it.second.specialization_info.get();
     }
+
+    if (stage == VK_SHADER_STAGE_COMPUTE_BIT &&
+        it.second.required_subgroup_size > 0)
+    {
```
Comment (dj2) [inline src/vulkan/engine_vulkan.cc:390]: Please run the formatter.

File: docs/amber_script.md [hunk 20]
```diff
@@ -435,6 +438,27 @@ The following commands are all specified within the `PIPELINE` command.
   FRAMEBUFFER_SIZE _width_ _height_
 ```
 
+```groovy
+  # Require that subgroups must be launched with all invocations active
+  # for given shader. Only valid in the compute stage.
+  REQUIRE_FULL_SUBGROUPS {name_of_shader}
```
Comment (dj2) [inline docs/amber_script.md:444]: Not a huge fan of this syntax for setting subgroup information. There seems to be 3 related things begin set here. So, a few questions. If these are always related to a given shader, shoudl these all be options on the `ATTACH` command?

Would this be better as a `SUBGROUP` block similar to `DEPTH` and `STENCIL`.

File: docs/amber_script.md [hunk 21]
```diff
@@ -435,6 +438,27 @@ The following commands are all specified within the `PIPELINE` command.
...
+  # Allow SubgroupSize to vary for given shader.
+  VARYING_SUBGROUP_SIZE {name_of_shader}
+```
+
+```groovy
+  # Require a specific SubgroupSize the for given shader:
+  # |subgroup_size| can be set one of the values below:
+  #  - a power-of-two integer that _must_ be greater or equal to minSubgroupSize
+  #    and be less than or equal to maxSubgroupSize
+  # - "MIN" to set the required subgroup size to the minSubgroupSize
```
Comment (dj2) [inline docs/amber_script.md:457]: Remove the `""`s around MIN and MAX

File: docs/amber_script.md (hunk 20 above)
Comment (dj2) [reply docs/amber_script.md:444]: Is this a per-shader item or a per-pipeline item, or shader/pipeline? Would you have one shader, in two pipelines with different subgroup information?

Comment (mmerecki) [reply docs/amber_script.md:444]: It's a per-shader/pipeline item. It is valid to use the same shader in two pipelines with different subgroup settings.

Comment (alan-baker) [reply docs/amber_script.md:444]: That syntax looks pretty good. I need to look at the details.

File: samples/config_helper_vulkan.cc (hunk 1 above)
Comment (mmerecki) [reply samples/config_helper_vulkan.cc]: removed

File: samples/config_helper_vulkan.cc (hunk 2 above)
Comment (mmerecki) [reply samples/config_helper_vulkan.cc]: done

File: samples/config_helper_vulkan.cc (hunk 3 above)
Comment (mmerecki) [reply samples/config_helper_vulkan.cc]: changed to a bool

File: samples/config_helper_vulkan.cc (hunk 4 above)
Comment (mmerecki) [reply samples/config_helper_vulkan.cc]: I run `clang-format`, should look ok now.

File: src/pipeline.cc (hunk 5 above)
Comment (mmerecki) [reply src/pipeline.cc]: Changed to an enum + numeric value.

File: src/vulkan/device.cc (hunk 17 above)
Comment (mmerecki) [reply src/vulkan/device.cc]: done

File: src/vulkan/engine_vulkan.cc (hunk 18 above)
Comment (mmerecki) [reply src/vulkan/engine_vulkan.cc]: done

File: src/vulkan/engine_vulkan.cc (hunk 19 above)
Comment (mmerecki) [reply src/vulkan/engine_vulkan.cc]: done

File: docs/amber_script.md (hunk 21 above)
Comment (mmerecki) [reply docs/amber_script.md]: I updated the syntax to match the form proposed in https://github.com/google/amber/pull/852#discussion_r418064490

File: src/amberscript/parser_subgroup_size_control_test.cc (hunk 6 above)
Comment (mmerecki) [reply src/amberscript/parser_subgroup_size_control_test.cc]: changed the test to use 256

File: src/pipeline.cc (hunk 7 above)
Comment (mmerecki) [reply src/pipeline.cc]: fixed

File: src/pipeline.h (hunk 8 above)
Comment (mmerecki) [reply src/pipeline.h]: fixed

File: src/script.h (hunk 9 above)
Comment (mmerecki) [reply src/script.h]: fixed

File: src/script.h (hunk 10 above)
Comment (mmerecki) [reply src/script.h]: changed to `begin`/`end`

File: src/vulkan/device.cc (hunk 11 above)
Comment (mmerecki) [reply src/vulkan/device.cc]: I removed this change completely.

File: src/vulkan/device.cc (hunk 12 above)
Comment (mmerecki) [reply src/vulkan/device.cc]: I removed this change completely.

File: src/vulkan/device.cc (hunk 14 above)
Comment (mmerecki) [reply src/vulkan/device.cc]: done

File: src/vulkan/device.cc (hunk 15 above)
Comment (mmerecki) [reply src/vulkan/device.cc]: This code is not querying features but device properties.

File: src/vulkan/device.cc (hunk 16 above)
Comment (mmerecki) [reply src/vulkan/device.cc]: fixed

File: src/vulkan/device.cc (hunk 17 above)
Comment (mmerecki) [reply src/vulkan/device.cc]: done

File: docs/amber_script.md (hunk 20 above)
Comment (mmerecki) [reply docs/amber_script.md]: I updated the syntax to match the form proposed in #852 (comment)

**Comment by dj2** [comment]
LGTM, I'm good when the bots are happy.

**Comment by mmerecki** [comment]
Is there a plan to switch to Android NDK r21 for Android builds? 

Vulkan header that is in r20 does not have vk_ext_subgroup_size_control.

**Comment by alan-baker** [comment]
@mmerecki, you can modify the `kokoro/ndk-build/build.sh` to grab NDK r21. There's no particular reason we're using r20.

**Comment by dj2** [comment]
And rebase to master if possible please.

File: tests/run_tests.py [hunk 22]
```diff
@@ -84,6 +84,11 @@
   "opencl_read_and_write_image3d_rgba32i.amber",
   "opencl_write_image.amber",
   "glsl_read_and_write_image3d_rgba32i.amber",
+   "compute_require_full_subgroups.amber",
```
Comment (dj2) [inline tests/run_tests.py:87]: Can you add a comment for why these are disabled?

Comment (mmerecki) [reply tests/run_tests.py]: I removed changes in this file.

File: kokoro/ndk-build/build.sh [hunk 23]
```diff
@@ -28,8 +28,8 @@ unzip -q ninja-linux.zip
 export PATH="$PWD:$PATH"
 
 # Get Android NDK.
```
Comment (alan-baker) [inline kokoro/ndk-build/build.sh:30]: This is still set to r20.

Comment (alan-baker) [reply kokoro/ndk-build/build.sh:30]: ANDROID_NDK env var that is.

**Comment by mmerecki** [comment]
I don't understand the latest CI-check-format failure. Is this an env issue?

**Comment by alan-baker** [comment]
The bot is unable to retrieve the clang release. Unrelated to your PR.

**Comment by dj2** [comment]
If the rest of the bots pass, we can ignore the format failure for now, can always fix that in a subsequent Cl.
